packages/language-server/src/plugins/typescript/TypeScriptPlugin.ts

```diff
@@ -78,74 +75,93 @@ export class TypeScriptPlugin
         this.configManager = configManager;
     }
 
-    getDiagnostics(document: Document): Diagnostic[] {
+    async getDiagnostics(document: Document): Promise<Diagnostic[]> {
         if (!this.featureEnabled('diagnostics')) {
             return [];
         }
 
         const { lang, tsDoc } = this.getLSAndTSDoc(document);
-        const isTypescript =
-            getScriptKindFromAttributes(tsDoc.getAttributes()) === ts.ScriptKind.TS;
+        const isTypescript = tsDoc.scriptKind === ts.ScriptKind.TSX;
+
+        // Document preprocessing failed, show parser error instead
+        if (tsDoc.parserError) {
+            return [
+                {
+                    range: tsDoc.parserError.range,
+                    severity: DiagnosticSeverity.Error,
+                    source: isTypescript ? 'ts' : 'js',
+                    message: tsDoc.parserError.message,
+                    code: tsDoc.parserError.code,
+                },
+            ];
+        }
 
         const diagnostics: ts.Diagnostic[] = [
-            ...lang.getSyntacticDiagnostics(tsDoc.getFilePath()!),
-            ...lang.getSuggestionDiagnostics(tsDoc.getFilePath()!),
-            ...lang.getSemanticDiagnostics(tsDoc.getFilePath()!),
+            ...lang.getSyntacticDiagnostics(tsDoc.filePath),
+            ...lang.getSuggestionDiagnostics(tsDoc.filePath),
+            ...lang.getSemanticDiagnostics(tsDoc.filePath),
         ];
 
+        const fragment = await tsDoc.getFragment();
+
         return diagnostics
-            .map(diagnostic => ({
+            .map((diagnostic) => ({
                 range: convertRange(tsDoc, diagnostic),
                 severity: mapSeverity(diagnostic.category),
                 source: isTypescript ? 'ts' : 'js',
                 message: ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n'),
                 code: diagnostic.code,
             }))
-            .map(diagnostic => mapDiagnosticToParent(tsDoc, diagnostic));
+            .map((diagnostic) => mapDiagnosticToParent(fragment, diagnostic))
+            .filter(
+                (diagnostic) => diagnostic.range.start.line >= 0 && diagnostic.range.end.line >= 0,
+            );
```

Comment:
I wonder if this could accidentally remove diagnostics which come from setup around the specific file? It seems unlikely, but at least worth bringing up

Reply:
I added this because in some rare cases the diagnostic could not be mapped correctly. But VSCode or the LSP (don't know which one) then seems to swallow all other errors aswell as soon as it encounters one diagnostic with negative lines. I will add a comment.